File: packages/tools/src/cli/commands/release.ts
import { buildCommand } from "@stricli/core";
import {
  performRelease,
  type ReleaseType,
  releaseTypeSchema,
} from "../../lib/release.ts";
import { parseWithZod } from "../../lib/zod.ts";

function createReleaseLogger() {
  const prefix = "[release] ";

  return {
    info: (message: string) => console.log(`${prefix}${message}`),
    start: (message: string) => console.log(`${prefix}▶ ${message}`),
    success: (message: string) => console.log(`${prefix}✔ ${message}`),
  };
}

type ReleaseArgs = [releaseType: ReleaseType];
type ReleaseCommandFlags = {
  dryRun: boolean;
};

export async function runReleaseCommand(
  flags: ReleaseCommandFlags,
  releaseType: ReleaseType,
): Promise<void> {
  const logger = createReleaseLogger();

  const result = await performRelease({
    cwd: process.cwd(),
    dryRun: flags.dryRun,
    logger,
    releaseType,
  });

  if (result.dryRun) {
    logger.success(
      `Dry run: would release ${result.version} from ${result.previousTag} to ${result.tag}`,
    );
    return;
  }

  logger.success(
    `Released ${result.version} from ${result.previousTag} to ${result.tag}`,
  );
}

export const releaseCommand = buildCommand<ReleaseCommandFlags, ReleaseArgs>({
  parameters: {
    flags: {
      dryRun: {
        kind: "boolean",
        brief:
          "Validate and show the next release without changing files or git state",
        default: false,
      },
    },
    positional: {
      kind: "tuple",
      parameters: [
        {
          brief: "Semver bump type",
          parse: async (input) =>
            await parseWithZod(input, {
              label: "release-type",
              schema: releaseTypeSchema,
            }),
        },
      ],
    },
    aliases: {
      n: "dryRun",
    },
  },
  docs: {
    brief: "Release the CLI package",
    fullDescription:
      "Bump release versions from current package versions and create a commit and signed tag.",
  },
  func: runReleaseCommand,
});

File: packages/tools/lib/src/commands/release.dart
import 'dart:io';

import 'package:proxer_tools/src/lib/args.dart';
import 'package:proxer_tools/src/lib/error.dart';
import 'package:proxer_tools/src/lib/release.dart';
import 'package:proxer_tools/src/lib/version.dart';

void _log(String message) {
  stdout.writeln('[release] $message');
}

Future<int> runReleaseCommand(List<String> args) async {
  final parsed = parseArgs(
    args,
    booleanFlags: const {'dry-run'},
    aliases: const {'n': 'dry-run'},
  );

  if (parsed.positionals.length != 1) {
    throw const ToolException('Usage: release <patch|minor|major> [--dry-run]');
  }

  final releaseType = parseReleaseType(parsed.positionals.single);
  final logger = ReleaseLogger(
    info: _log,
    start: (message) => _log('▶ $message'),
  );

  final result = await performRelease(
    cwd: Directory.current.path,
    dryRun: parsed.flag('dry-run'),
    logger: logger,
    releaseType: releaseType,
  );

  if (result.dryRun) {
    _log(
      '✔ Dry run: would release ${result.version} '
      'from ${result.previousTag} to ${result.tag}',
    );

    return 0;
  }

  _log(
    '✔ Released ${result.version} '
    'from ${result.previousTag} to ${result.tag}',
  );

  return 0;
}
